### PR TITLE
ARROW-347: Add method to pass CallBack to get Transfer Pair

### PR DESCRIPTION
**Pull request by StevenMPhillips**

### CODE REVIEW

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 1]
```diff
@@ -235,12 +240,17 @@ public Field getField() {
...
   @Override
   public TransferPair getTransferPair(BufferAllocator allocator) {
-    return new TransferImpl(name, allocator);
+    return getTransferPair(name, allocator);
   }
 
   @Override
   public TransferPair getTransferPair(String ref, BufferAllocator allocator) {
-    return new TransferImpl(ref, allocator);
+    return getTransferPair(ref, allocator, null);
```
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionVector.java]: why do we need to keep this method?

File: java/vector/src/main/java/org/apache/arrow/vector/BaseDataValueVector.java [hunk 2]
```diff
@@ -76,6 +78,11 @@ public void close() {
   }
 
   @Override
+  public TransferPair getTransferPair(String ref, BufferAllocator allocator, CallBack callBack) {
+    return getTransferPair(ref, allocator);
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/BaseDataValueVector.java]: why do we drop the callback here?

File: java/vector/src/main/java/org/apache/arrow/vector/ValueVector.java [hunk 3]
```diff
@@ -106,6 +107,8 @@
 
   TransferPair getTransferPair(String ref, BufferAllocator allocator);
 
+  TransferPair getTransferPair(String ref, BufferAllocator allocator, CallBack callBack);
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/ValueVector.java]: why don't we just add a parameter to the existing method?
We should avoid duplicating those methods unless necessary.
Please add javadoc to interface/abstract methods

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java [hunk 4]
```diff
@@ -135,7 +135,12 @@ public FieldVector getDataVector() {
 
   @Override
   public TransferPair getTransferPair(String ref, BufferAllocator allocator) {
-    return new TransferImpl(ref, allocator);
+    return getTransferPair(ref, allocator, null);
+  }
+
+  @Override
+  public TransferPair getTransferPair(String ref, BufferAllocator allocator, CallBack callBack) {
+    return null;
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: ? that's different from the original behavior. What's the intent?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/NullableMapVector.java [hunk 5]
```diff
@@ -93,6 +93,11 @@ public TransferPair makeTransferPair(ValueVector to) {
 
   @Override
   public TransferPair getTransferPair(String ref, BufferAllocator allocator) {
+    return new NullableMapTransferPair(this, new NullableMapVector(ref, allocator, null), false);
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/NullableMapVector.java]: could just call the other one with null param

File: java/vector/src/main/java/org/apache/arrow/vector/complex/NullableMapVector.java [hunk 6]
```diff
@@ -83,7 +83,7 @@ public FieldReader getReader() {
 
   @Override
   public TransferPair getTransferPair(BufferAllocator allocator) {
-    return new NullableMapTransferPair(this, new NullableMapVector(name, allocator, callBack), false);
+    return new NullableMapTransferPair(this, new NullableMapVector(name, allocator, null), false);
```
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/NullableMapVector.java]: could just call the one bellow with name as ref.

**Review by julienledem — CHANGES_REQUESTED** [review]
Hi @StevenMPhillips 
See my comments bellow

**Comment by wesm** [comment]
@StevenMPhillips ping, also needs rebase

**Comment by wesm** [comment]
ping, can we close out all the Java PRs in advance of Arrow 0.2?

**Comment by wesm** [comment]
Pinging again on this

**Comment by julienledem** [comment]
@StevenMPhillips can you rebase and respond to comments?

**Comment by wesm** [comment]
pinging on this one

**Pull request closed without merging**